### PR TITLE
Start hand; deal cards

### PR DESCRIPTION
**Pull request by jonah-katz**

### CODE REVIEW

File: nodejs/src/classes/table.ts [hunk 1]
```diff
@@ -1,78 +1,230 @@
 import Deck from './deck';
 
-export type Player = {
-  user_id: string;
+type UserId = string;
+
+// Member = someone sitting at table
+// (but necessarily playing. only status=active players play)
```
Comment (jacklieblich) [inline nodejs/src/classes/table.ts:6]:
```suggestion
// (but not necessarily playing. only status=active players play)
```

File: nodejs/src/classes/table.ts [hunk 2]
```diff
@@ -1,78 +1,230 @@
...
+  addMember(member: Member) {
+    this.members.push(member);
+  }
+
+  deal() {
+    let deck = new Deck();
+    this.hand =  {
+      dealer: this.members.filter(p => p.status === 'active')[0].user_id,
+      deck: deck,
+      players: this.members.filter(p => p.status === 'active').map(p => ({
```
Comment (jacklieblich) [inline nodejs/src/classes/table.ts:89]:
```suggestion
    const activeMembers = this.members.filter(p => p.status === 'active')
    this.hand =  {
      dealer: activeMembers[0].user_id,
      deck: deck,
      players: activeMembers.map(p => ({
```

File: nodejs/src/classes/table.ts [hunk 3]
```diff
@@ -1,78 +1,230 @@
...
+      betting_rounds: [[]]
+    }
+
+    // do blinds
+    let index_of_dealer = -1;
+    this.hand.players.forEach((a, i) => {
+      if (a.user_id === this.hand!.dealer) {
+        index_of_dealer = i;
+      }
+    });
```
Comment (jacklieblich) [inline nodejs/src/classes/table.ts:104]:
```suggestion
    let index_of_dealer = this.hand.players.findIndex((player) => player.user_id === this.hand!.dealer)
```

File: nodejs/src/classes/table.ts [hunk 4]
```diff
@@ -1,78 +1,230 @@
...
+  }
+
+  addMember(member: Member) {
+    this.members.push(member);
+  }
+
+  deal() {
+    let deck = new Deck();
+    this.hand =  {
+      dealer: this.members.filter(p => p.status === 'active')[0].user_id,
```
Comment (jacklieblich) [inline nodejs/src/classes/table.ts:87]: How does the dealer rotate?

File: nodejs/src/classes/table.ts [hunk 5]
```diff
@@ -1,78 +1,230 @@
...
+    // do blinds
+    let index_of_dealer = -1;
+    this.hand.players.forEach((a, i) => {
+      if (a.user_id === this.hand!.dealer) {
+        index_of_dealer = i;
+      }
+    });
+
+    let small_blind_index = index_of_dealer + 1;
+    let big_blind_index = small_blind_index + 1;
```
Comment (jacklieblich) [inline nodejs/src/classes/table.ts:107]: What about heads-up?

from [wikipedia](https://en.wikipedia.org/wiki/Texas_hold_%27em#Betting_structures)

> When only two players remain, special 'head-to-head' or 'heads up' rules are enforced and the blinds are posted differently. In this case, the person with the dealer button posts the small blind, while his/her opponent places the big blind.

File: nodejs/src/classes/table.ts [hunk 6]
```diff
@@ -1,78 +1,230 @@
...
+      players: this.members.filter(p => p.status === 'active').map(p => ({
+        user_id: p.user_id,
+        card_one: deck.pop1(),
+        card_two: deck.pop1(),
+      })),
+      community_cards: [],
+      betting_rounds: [[]]
+    }
+
+    // do blinds
```
Comment (jacklieblich) [inline nodejs/src/classes/table.ts:98]: Would be sick to pull this all out into it's own `postBlinds` function.

File: nodejs/src/classes/deck.ts [hunk 7]
```diff
@@ -2,6 +2,38 @@ export default class Deck {
...
+
+    return array;
+  }
+
+  pop1(): number {
+    return this.cards.pop()!;
+  }
+
+  pop3(): Array<number>  {
+    return [this.cards.pop()!, this.cards.pop()!, this.cards.pop()!];
```
[the remark this answers is not in the document]
Comment (jonah-katz) [reply nodejs/src/classes/deck.ts:37]: Correct. Value could potentially be undefined. Typescript thing

File: nodejs/src/classes/table.ts (hunk 4 above)
Comment (jonah-katz) [reply nodejs/src/classes/table.ts:87]: The first dealer is the first active member who joined. 

Later, we'll just go to the next active person I think? The members array maintains the order the members join the table

File: nodejs/src/classes/table.ts (hunk 5 above)
Comment (jonah-katz) [reply nodejs/src/classes/table.ts:107]: Ah shit

File: nodejs/src/classes/table.ts (hunk 6 above)
Comment (jonah-katz) [reply nodejs/src/classes/table.ts:98]: Agree.. especially now that there's head to head... will do that later

**Pull request merged**